File: src/frame/modes.c
// SPDX-License-Identifier: Apache-2.0
#include "vgacap/frame.h"

static const vgaframe_mode_t table[] = {
    { "640x480@60",  640, 16,  96,  48, 480, 10, 2, 33, 0, 0 },
    { "640x480@72",  640, 24,  40, 128, 480,  9, 3, 28, 0, 0 },
    { "640x480@75",  640, 16,  64, 120, 480,  1, 3, 16, 0, 0 },
    { "720x400@70",  720, 18, 108,  54, 400, 12, 2, 35, 0, 1 },
    { "800x600@56",  800, 24,  72, 128, 600,  1, 2, 22, 1, 1 },
    { "800x600@60",  800, 40, 128,  88, 600,  1, 4, 23, 1, 1 },
    { "1024x768@60", 1024, 24, 136, 160, 768,  3, 6, 29, 0, 0 },
};

const vgaframe_mode_t *vgaframe_modes(size_t *count) {
    if (count) *count = sizeof table / sizeof table[0];
    return table;
}

const vgaframe_mode_t *vgaframe_mode_match(uint32_t clocks_per_line, uint32_t lines_per_frame) {
    for (size_t i = 0; i < sizeof table / sizeof table[0]; i++) {
        const vgaframe_mode_t *m = &table[i];
        uint32_t cpl = (uint32_t)m->h_active + m->h_front + m->h_sync + m->h_back;
        uint32_t lpf = (uint32_t)m->v_active + m->v_front + m->v_sync + m->v_back;
        if (cpl == clocks_per_line && lpf == lines_per_frame) return m;
    }
    return NULL;
}

File: python/vgacap/modes.py
# SPDX-License-Identifier: Apache-2.0
"""Mirror of the built-in mode table (see src/frame/modes.c)."""
from __future__ import annotations
from dataclasses import dataclass


@dataclass(frozen=True)
class Mode:
    name: str
    h_active: int
    h_front: int
    h_sync: int
    h_back: int
    v_active: int
    v_front: int
    v_sync: int
    v_back: int
    h_sync_positive: bool
    v_sync_positive: bool


_TABLE = [
    Mode("640x480@60",  640, 16,  96,  48, 480, 10, 2, 33, False, False),
    Mode("640x480@72",  640, 24,  40, 128, 480,  9, 3, 28, False, False),
    Mode("640x480@75",  640, 16,  64, 120, 480,  1, 3, 16, False, False),
    Mode("720x400@70",  720, 18, 108,  54, 400, 12, 2, 35, False, True),
    Mode("800x600@56",  800, 24,  72, 128, 600,  1, 2, 22, True,  True),
    Mode("800x600@60",  800, 40, 128,  88, 600,  1, 4, 23, True,  True),
    Mode("1024x768@60", 1024, 24, 136, 160, 768,  3, 6, 29, False, False),
]

MODES: dict[str, Mode] = {m.name: m for m in _TABLE}
